core/api/src/services/ledger/facade/onchain-reconcile.ts

```diff
@@ -0,0 +1,78 @@
+import { MainBook } from "../books"
+
+import { translateToLedgerTx } from ".."
+import { getBankOwnerWalletId } from "../caching"
+import { UnknownLedgerError } from "../domain/errors"
+import { persistAndReturnEntry } from "../helpers"
+import { FeeOnlyEntryBuilder } from "../domain/fee-only-entry-builder"
+
+import { staticAccountIds } from "./static-account-ids"
+
+import { LedgerTransactionType, toLiabilitiesWalletId } from "@/domain/ledger"
+import { AmountCalculator } from "@/domain/shared"
+
+const calc = AmountCalculator()
+
+export const recordReceiveOnChainFeeReconciliation = async ({
+  estimatedFee,
+  actualFee,
+  metadata,
+}: {
+  estimatedFee: BtcPaymentAmount
+  actualFee: BtcPaymentAmount
+  metadata: AddOnChainFeeReconciliationLedgerMetadata
+}) => {
+  const accountIds = await staticAccountIds()
+  if (accountIds instanceof Error) return accountIds
+
+  let entry = MainBook.entry("")
+  if (actualFee.amount > estimatedFee.amount) {
+    const btcFeeDifference = calc.sub(actualFee, estimatedFee)
+    const builder = FeeOnlyEntryBuilder({
+      staticAccountIds: accountIds,
+      entry,
+      metadata,
+      btcFee: btcFeeDifference,
+    })
+    entry = builder.debitBankOwner().creditOnChain()
+  } else {
+    const btcFeeDifference = calc.sub(estimatedFee, actualFee)
+    const builder = FeeOnlyEntryBuilder({
+      staticAccountIds: accountIds,
+      entry,
+      metadata,
+      btcFee: btcFeeDifference,
+    })
+    entry = builder.debitOnChain().creditBankOwner()
+  }
+
+  return persistAndReturnEntry({
+    entry,
+    hash: metadata.hash,
+  })
+}
+
+export const isOnChainFeeReconciliationTxn = (
+  txn: LedgerTransaction<WalletCurrency>,
+): boolean =>
+  txn.type === LedgerTransactionType.OnchainPayment && txn.address === undefined
```

Comment:
not sure when we began to use non undefinied/null/empty validation (should be `!txn.address` we already had an issue in prod with this) also.. should not include bankowner wallet id validation?

Reply:
In this case I was thinking it should specifically check for the absence of the `address` property (since the `OnChainAddress` can technically also be falsy/empty even though it should never be). Or, what sort of issue has this thrown in prod?

> should not include bankowner wallet id validation
Maybe not necessary since the reconciliation transaction only ever involves `Liabilities:bankOwner` account and `Asset:OnChain` account and I believe we only bring in `Liabilities` accounts in our ledger `get` methods. `address` absence should be enough for this check?

Also for this check, ideally it should be a different tx `type` value instead of also being an `onchain_payment` so we don't even need the check, but that might be a tricky migration to do. And I also considered maybe setting a custom description for these types of transactions but then there was no way to guarantee that someone couldn't set the same description on normal onchain-payment transactions which would break the heuristic. I switched from checking for `satsFee` presence to `address` presence since this is more isolated and might less likely eventually be populated on the reconciliation transactions.

Reply:
at least with validations in data from external services we must be more "defensive", we cant guarantee the undefined so it is safer just to use `!txn.address`

FYI: the error was a validation using undefined and a service returning null value (typescript in theory was ok but you know we cant trust TS )

Reply:
Oh yea ok that makes sense.

For here though, `address` is set explicitly in our `translateToLedgerTx` function, so there's also less risk of it being something unexpected? We handle the uncertain-falsy-check-from-external-service in the line `tx.payee_addresses && tx.payee_addresses.length > 0`